File: Assets/Scripts/Tile.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class Tile : MonoBehaviour {
   
    public Vector2 GridPosition = Vector2.zero;
    List<Tile> neighbours = new List<Tile>();
    public int movementCost = 1;
    public bool imPassable = false;

    private Renderer rend;
    private Color defaultColor;

    // Use this for initialization
	void Start () {
        rend = GetComponent<Renderer>();
        defaultColor = rend.material.color;
        generateNeighbours();
    }
	
	// Update is called once per frame
	void Update () {
	
	}

    void OnMouseEnter()
    {
        if (GameManager.Instance.players[GameManager.Instance.PlayerIndex].moving)
        {
            rend.material.color = Color.blue;
        }
        else if (GameManager.Instance.players[GameManager.Instance.PlayerIndex].attacking)
        {
            rend.material.color = Color.red;
        }
    }
    void OnMouseExit()
    {
        rend.material.color = defaultColor;
    }
    void OnMouseDown()
    {
        if (GameManager.Instance.players[GameManager.Instance.PlayerIndex].moving)
        {
            GameManager.Instance.moveCurrentPlayer(this);
        }
        else if (GameManager.Instance.players[GameManager.Instance.PlayerIndex].attacking)
        {
            GameManager.Instance.attackWithCurrentPlayer(this);
        }
        else // added temporarly to create imPassable areas for pathfinding testing
        {
            imPassable = !imPassable;
            if (imPassable)
            {
                Rend.material.color = Color.black;
            }
            else
            {
                Rend.material.color = Color.white;
            }
        }
    }

    void GenerateNeighbours()
    {
        neighbours = new ListTile();

        // y+ (up)
        if (GridPosition.y > 0)
        {
            Vector2 n = new Vector2(GridPosition.x, GridPosition.y - 1);
            neighbours.Add(GameManager.Instance.map[(int)n.x][(int)n.y]);
        }

        // y- (down)
        if (GridPosition.y < GameManager.Instance.map.Count - 1)
        {
            Vector2 n = new Vector2(GridPosition.x, GridPosition.y + 1);
            neighbours.Add(GameManager.Instance.map[(int)n.x][(int)n.y]);
        }

        // x+ (right)
        if (GridPosition.x > 0)
        {
            Vector2 n = new Vector2(GridPosition.x - 1, GridPosition.y);
            neighbours.Add(GameManager.Instance.map[(int)n.x][(int)n.y]);
        }

        // x- (left)
        if (GridPosition.x < GameManager.Instance.map.Count - 1)
        {
            Vector2 n = new Vector2(GridPosition.x + 1, GridPosition.y);
            neighbours.Add(GameManager.Instance.map[(int)n.x][(int)n.y]);
        }
    }
}

File: Assets/Scripts/Tile.py
import pygame
from pygame.math import Vector2

class Tile:
    def __init__(self, position=(0, 0), size=50):
        self.position = Vector2(position)
        self.grid_position = Vector2(0, 0)
        self.size = size
        self.color = (255, 255, 255)  # Default color (white)
        self.default_color = self.color
        self.im_passable = False

    def render(self, screen):
        """Render the tile as a rectangle on the screen."""
        rect = pygame.Rect(self.position.x, self.position.y, self.size, self.size)
        pygame.draw.rect(screen, self.color, rect)
        pygame.draw.rect(screen, (0, 0, 0), rect, 1)  # Draw border

    def on_mouse_enter(self):
        """Change the tile's color when the mouse hovers over it."""
        self.color = (200, 200, 200)  # Light gray

    def on_mouse_exit(self):
        """Revert the tile's color when the mouse leaves."""
        self.color = self.default_color

    def on_mouse_down(self, current_player):
        """Handle mouse click events on the tile."""
        if current_player.moving:
            current_player.move_to(self)
        elif current_player.attacking:
            current_player.attack(self)
        else:
            # Toggle impassable state for testing
            self.im_passable = not self.im_passable
            self.color = (0, 0, 0) if self.im_passable else self.default_color